include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -840,6 +840,173 @@ class __static_monotonic_dispatcher<::std::integer_sequence<::std::uint16_t, _X,
     }
 };
 
+struct __scalar_load_op
+{
+    oneapi::dpl::__internal::__pstl_assign __assigner;
+    template <typename _IdxType1, typename _IdxType2, typename _SourceAcc, typename _DestAcc>
+    void
+    operator()(_IdxType1 __idx_source, _IdxType2 __idx_dest, _SourceAcc __source_acc, _DestAcc __dest_acc) const
+    {
+        __assigner(__source_acc[__idx_source], __dest_acc[__idx_dest]);
+    }
+};
+
+template <std::uint8_t __vec_size>
+struct __vector_load
+{
+    static_assert(__vec_size <= 4, "Only vector sizes of 4 or less are supported");
+    std::size_t __full_range_size;
+    template <typename _IdxType, typename _LoadOp, typename... _Rngs>
+    void
+    operator()(/*__is_full*/ std::true_type, _IdxType __start_idx, _LoadOp __load_op, _Rngs&&... __rngs) const
+    {
+        _ONEDPL_PRAGMA_UNROLL
+        for (std::uint8_t __i = 0; __i < __vec_size; ++__i)
+            __load_op(__start_idx + __i, __i, __rngs...);
+    }
+
+    template <typename _IdxType, typename _LoadOp, typename... _Rngs>
+    void
+    operator()(/*__is_full*/ std::false_type, _IdxType __start_idx, _LoadOp __load_op, _Rngs&&... __rngs) const
+    {
+        std::uint8_t __elements = std::min(std::size_t{__vec_size}, std::size_t{__full_range_size - __start_idx});
+        for (std::uint8_t __i = 0; __i < __elements; ++__i)
+            __load_op(__start_idx + __i, __i, __rngs...);
+    }
+};
+
+template <typename _TransformOp>
+struct __scalar_store_transform_op
+{
+    _TransformOp __transform;
+    // Unary transformations into an output buffer
+    template <typename _IdxType1, typename _IdxType2, typename _SourceAcc, typename _DestAcc>
+    void
+    operator()(_IdxType1 __idx_source, _IdxType2 __idx_dest, _SourceAcc __source_acc, _DestAcc __dest_acc) const
+    {
+        __transform(__source_acc[__idx_source], __dest_acc[__idx_dest]);
+    }
+    // Binary transformations into an output buffer
+    template <typename _IdxType1, typename _IdxType2, typename _Source1Acc, typename _Source2Acc, typename _DestAcc>
+    void
+    operator()(_IdxType1 __idx_source, _IdxType2 __idx_dest, _Source1Acc __source1_acc, _Source2Acc __source2_acc,
+               _DestAcc __dest_acc) const
+    {
+        __transform(__source1_acc[__idx_source], __source2_acc[__idx_source], __dest_acc[__idx_dest]);
+    }
+};
+
+// TODO: Consider unifying the implementations of __vector_walk, __vector_load, __vector_store, and potentially
+// __strided_loop with some common, generic utility
+template <std::uint8_t __vec_size>
+struct __vector_walk
```

Comment:
This struct is very similar to `__vector_load`... and I have the same questions - https://github.com/uxlfoundation/oneDPL/pull/1976/files#r1933883494